### PR TITLE
Rewrite patching logic to use a Thrift service for request proxying.

### PR DESCRIPTION
**Pull request by abhinav**
The VCR functionality is implemented by a Thrift service called VCRProxy.

Also,

* Instead of patching `PeerGroup.choose`, we now patch `TChannel.request`.
* Added support for matcher customization for cassettes.

@breerly @blampe @junchaowu

### CODE REVIEW

**Comment by abhinav** [comment]
FYI I mean to merge the vcr branch into master after this is reviewed. I think it is ready for use. We can address the TODOs and parsing of request/response bodies later.

File: tchannel/testing/vcr/__init__.py [hunk 1]
```diff
@@ -22,8 +22,8 @@
 VCR
```
Comment (blampe) [inline tchannel/testing/vcr/__init__.py]: Oh boo you put this all in __init__.

**Comment by blampe** [comment]
LGTM!

File: tchannel/testing/vcr/__init__.py (hunk 1 above)
Comment (breerly) [inline tchannel/testing/vcr/__init__.py]: @abhinav please move out and instead import into init

Comment (abhinav) [inline tchannel/testing/vcr/__init__.py]: fixed

**Pull request merged**